File: src/main/kotlin/dev/murad/shipping/entity/models/insert/RingsInsertBargeModel.java
package dev.murad.shipping.entity.models.insert;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import dev.murad.shipping.HumVeeMod;
import dev.murad.shipping.entity.Colorable;
import net.minecraft.client.model.EntityModel;
import net.minecraft.client.model.geom.ModelLayerLocation;
import net.minecraft.client.model.geom.ModelPart;
import net.minecraft.client.model.geom.PartPose;
import net.minecraft.client.model.geom.builders.*;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;
import org.jetbrains.annotations.NotNull;

public class RingsInsertBargeModel<T extends Entity & Colorable> extends EntityModel<T> {
    // This layer location should be baked with EntityRendererProvider.Context in the entity renderer and passed into this model's constructor
    public static final ModelLayerLocation LAYER_LOCATION = new ModelLayerLocation(ResourceLocation.tryBuild(HumVeeMod.MOD_ID, "rings_insert_barge_model"), "main");
    private final ModelPart bb_main;

    public RingsInsertBargeModel(ModelPart root) {
        this.bb_main = root.getChild("bb_main");
    }

    public static LayerDefinition createBodyLayer() {
        MeshDefinition meshdefinition = new MeshDefinition();
        PartDefinition main = meshdefinition.getRoot()
                .addOrReplaceChild("bb_main",
                        CubeListBuilder.create()
                                .texOffs(0, 37).addBox(-2.0F, -33.0F, -2.0F, 4.0F, 4.0F, 4.0F),
                        PartPose.offset(0.0F, 23.0F, 0.0F));

        main.addOrReplaceChild("ring", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0F, -29.0F, -5.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(36, 31).addBox(-5.0F, -29.0F, 3.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(10, 41).addBox(3.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F)
                        .texOffs(38, 4).addBox(-5.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F),
                PartPose.offset(0.0F, 0.0F, 0.0F));

        main.addOrReplaceChild("ring2", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0F, -29.0F, -5.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(36, 31).addBox(-5.0F, -29.0F, 3.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(10, 41).addBox(3.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F)
                        .texOffs(38, 4).addBox(-5.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F),
                PartPose.offset(0.0F, -3.0F, 0.0F));

        main.addOrReplaceChild("ring3", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0F, -29.0F, -5.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(36, 31).addBox(-5.0F, -29.0F, 3.0F, 10.0F, 2.0F, 2.0F)
                        .texOffs(10, 41).addBox(3.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F)
                        .texOffs(38, 4).addBox(-5.0F, -29.0F, -3.0F, 2.0F, 2.0F, 6.0F),
                PartPose.offset(0.0F, -7.0F, 0.0F));

        return LayerDefinition.create(meshdefinition, 64, 64);
    }

    @Override
    public void setupAnim(T entity, float limbSwing, float limbSwingAmount, float ageInTicks, float netHeadYaw, float headPitch) {

    }

	@Override
	public void renderToBuffer(@NotNull PoseStack pPoseStack, @NotNull VertexConsumer pBuffer, int pPackedLight, int pPackedOverlay, int pColor) {
		bb_main.render(pPoseStack, pBuffer, pPackedLight, pPackedOverlay);
	}
}

File: src/main/kotlin/com/github/bonndan/humblevehicles/entity/models/insert/RingsInsertBargeModel.kt
package com.github.bonndan.humblevehicles.entity.models.insert

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import com.github.bonndan.humblevehicles.HumVeeMod
import com.github.bonndan.humblevehicles.entity.Colorable
import net.minecraft.client.model.EntityModel
import net.minecraft.client.model.geom.ModelLayerLocation
import net.minecraft.client.model.geom.ModelPart
import net.minecraft.client.model.geom.PartPose
import net.minecraft.client.model.geom.builders.CubeListBuilder
import net.minecraft.client.model.geom.builders.LayerDefinition
import net.minecraft.client.model.geom.builders.MeshDefinition
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity

class RingsInsertBargeModel<T>(root: ModelPart) : EntityModel<T>() where T : Entity, T : Colorable {

    private val bb_main: ModelPart

    init {
        this.bb_main = root.getChild("bb_main")
    }

    override fun setupAnim(
        entity: T?,
        limbSwing: Float,
        limbSwingAmount: Float,
        ageInTicks: Float,
        netHeadYaw: Float,
        headPitch: Float
    ) {
    }

    override fun renderToBuffer(
        pPoseStack: PoseStack,
        pBuffer: VertexConsumer,
        pPackedLight: Int,
        pPackedOverlay: Int,
        pColor: Int
    ) {
        bb_main.render(pPoseStack, pBuffer, pPackedLight, pPackedOverlay)
    }

    companion object {
        // This layer location should be baked with EntityRendererProvider.Context in the entity renderer and passed into this model's constructor
        val LAYER_LOCATION: ModelLayerLocation =
            ModelLayerLocation(ResourceLocation.tryBuild(HumVeeMod.MOD_ID, "rings_insert_barge_model"), "main")

        fun createBodyLayer(): LayerDefinition {
            val meshdefinition = MeshDefinition()
            val main = meshdefinition.getRoot()
                .addOrReplaceChild(
                    "bb_main",
                    CubeListBuilder.create()
                        .texOffs(0, 37).addBox(-2.0f, -33.0f, -2.0f, 4.0f, 4.0f, 4.0f),
                    PartPose.offset(0.0f, 23.0f, 0.0f)
                )

            main.addOrReplaceChild(
                "ring", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0f, -29.0f, -5.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(36, 31).addBox(-5.0f, -29.0f, 3.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(10, 41).addBox(3.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f)
                    .texOffs(38, 4).addBox(-5.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f),
                PartPose.offset(0.0f, 0.0f, 0.0f)
            )

            main.addOrReplaceChild(
                "ring2", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0f, -29.0f, -5.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(36, 31).addBox(-5.0f, -29.0f, 3.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(10, 41).addBox(3.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f)
                    .texOffs(38, 4).addBox(-5.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f),
                PartPose.offset(0.0f, -3.0f, 0.0f)
            )

            main.addOrReplaceChild(
                "ring3", CubeListBuilder.create().texOffs(38, 0).addBox(-5.0f, -29.0f, -5.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(36, 31).addBox(-5.0f, -29.0f, 3.0f, 10.0f, 2.0f, 2.0f)
                    .texOffs(10, 41).addBox(3.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f)
                    .texOffs(38, 4).addBox(-5.0f, -29.0f, -3.0f, 2.0f, 2.0f, 6.0f),
                PartPose.offset(0.0f, -7.0f, 0.0f)
            )

            return LayerDefinition.create(meshdefinition, 64, 64)
        }
    }
}